replication-api-impl/src/main/java/com/connexta/replication/api/impl/queue/memory/MemorySiteQueue.java

```diff
@@ -0,0 +1,426 @@
+import java.util.Deque;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.Nullable;
+
+/** Provides an in-memory implementation for a site specific queue. */
+// locking behavior is similar to LinkedBlockingQueue such that we can control capacity
+// and size across all internal queues and also to properly handle polling from multiple
+// different site queues when implementing the composite queue
+public class MemorySiteQueue implements MemoryQueue, SiteQueue {
+  private static final int MAX_PRIORITY = 9;
+  private static final int MIN_PRIORITY = 0;
+
+  private final MemoryQueueBroker broker;
+
+  private final String site;
+
+  /** List of all internal queues indexed by their corresponding priorities. */
+  private final List<Deque<MemoryTask>> pending = new ArrayList<>();
+
+  /** Ordered set of active tasks that are still locked and being worked on. */
+  private final Set<MemoryTask> active = new LinkedHashSet<>();
+
+  /** The capacity bound, or Integer.MAX_VALUE if none. */
+  private final int capacity;
+
+  /** Current number of tasks. */
+  private final AtomicInteger count;
+
+  /** Current number of pending tasks. */
+  private final AtomicInteger pendingCount;
+
+  /** Current number of active tasks. */
+  private final AtomicInteger activeCount;
+
+  /** Lock held by take, poll, etc. */
+  private final ReentrantLock takeLock = new ReentrantLock();
+
+  /** Wait queue for waiting takes. */
+  private final Condition notEmpty = takeLock.newCondition();
+
+  /** Lock held by put, offer, etc. */
+  private final ReentrantLock putLock = new ReentrantLock();
+
+  /** Wait queue for waiting puts. */
+  private final Condition notFull = putLock.newCondition();
+
+  private final MeterRegistry registry;
+
+  private final Counter queuedCounter;
+  private final Counter requeuedCounter;
+  private final Counter successCounter;
+  private final Counter failCounter;
+  private final Timer timer;
+  private final Timer pendingTimer;
+  private final Timer activeTimer;
+
+  /**
+   * Instantiates a default in-memory site queue.
+   *
+   * @param broker the broker for which we are instantiating this queue
+   * @param site the corresponding site id
+   * @param capacity the capacity of the queue this broker manages
+   * @param registry the micrometer registry to report metrics
+   */
+  public MemorySiteQueue(
+      MemoryQueueBroker broker, String site, int capacity, MeterRegistry registry) {
+    final Iterable<Tag> tags = Collections.singleton(Tag.of("site", site));
+
+    this.broker = broker;
+    this.site = site;
+    this.capacity = capacity;
+    this.registry = registry;
+    this.count = registry.gauge("replication_tasks_in_queue", tags, new AtomicInteger());
```

Comment:
✏️ We may want to rename these to a format like `replication_queue_*`, so `replication_queue_tasks`, `replication_queue_active_tasks`, etc since tasks is the unit of this measurement, it should be on the end. 

Also I just looked and we should separate with a `.` for optimal micrometer compatibility. My bad :)